### PR TITLE
add MuxLookup.fromEnum

### PR DESCRIPTION
**Pull request by albertchen-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### API Impact

adds `chisel3.util.MuxLookup.fromEnum`

#### Backend Code Generation Impact

none

#### Desired Merge Strategy

Squash

#### Release Notes

add `chisel3.util.MuxLookup.fromEnum`

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x` or `3.6.x` depending on impact, API modification or big change: `5.0.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/util/Mux.scala [hunk 1]
```diff
@@ -80,6 +80,17 @@ object MuxLookup {
...
   }
+
+  /** Same as MuxLookup.apply but for EnumTypes
+    *
+    * @param key a key to search for
+    * @param default a default value if nothing is found
+    * @param mapping a sequence to search of keys and values
+    * @return the value found or the default if not
+    */
+  def fromEnum[S <: EnumType, T <: Data](key: S, default: T, mapping: Seq[(S, T)]): T = {
```
Comment (ekiwi) [inline src/main/scala/chisel3/util/Mux.scala:91]: what happens if you try to overload the `apply` function? Does the Scala compiler complain?

**Comment by mmaloney-sf** [comment]
Thanks @albertchen-sifive!

Would you mind adding a short code example to how to use this new method in the docstring?

File: src/main/scala/chisel3/util/Mux.scala (hunk 1 above)
Comment (albertchen-sifive) [reply src/main/scala/chisel3/util/Mux.scala:91]: I get a compile error `method apply is defined twice`. https://scastie.scala-lang.org/eO18f45oQgKIi2GiBhMNYg

Comment (ekiwi) [reply src/main/scala/chisel3/util/Mux.scala:91]: OK. Sad :(

**Comment by mwachs5** [comment]
> I think I'll change the MuxLookup.fromEnum method to follow the suggested signature in the issue and then make a separate PR that adds a new version of MuxLookup.apply with the deprecation. Is that OK?

Either order is fine with me!

**Review by mwachs5 — APPROVED** [review]

File: src/main/scala/chisel3/util/Mux.scala (hunk 1 above)
Comment (jackkoenig) [reply src/main/scala/chisel3/util/Mux.scala]: We're successfully overloading apply now 🙂 see https://github.com/chipsalliance/chisel/pull/3095

**Pull request merged**